cirq/sim/clifford/act_on_stabilizer_ch_form_args.py

```diff
@@ -0,0 +1,95 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A protocol for implementing high performance clifford tableau evolutions
+ for Clifford Simulator."""
+
+from typing import Any, Iterable, TYPE_CHECKING
+
+import numpy as np
+
+from cirq.ops import common_gates
+from cirq.ops import pauli_gates
+from cirq.ops.clifford_gate import SingleQubitCliffordGate
+from cirq.protocols import has_unitary, num_qubits, unitary
+from cirq.sim.clifford.stabilizer_state_ch_form import StabilizerStateChForm
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class ActOnStabilizerCHFormArgs:
+    """State and context for an operation acting on a stabilizer state in CH
+    form. To act on this object, directly edit the `state` property, which is
+    storing the stabilizer state of the quantum system with one axis per qubit.
+    Measurements are currently not supported on this object.
+    """
+
+    def __init__(self, state: StabilizerStateChForm, axes: Iterable[int]):
+        """
+        Args:
+            state: The StabilizerStateChForm to act on. Operations are expected
+                to perform inplace edits of this object.
+            axes: The indices of axes corresponding to the qubits that the
+                operation is supposed to act upon.
+        """
+        self.state = state
+        self.axes = tuple(axes)
+
+    def _act_on_fallback_(self, action: Any, allow_decompose: bool):
+        strats = []
+        if allow_decompose:
+            strats.append(
+                _strat_act_on_stabilizer_ch_form_from_single_qubit_decompose)
+        for strat in strats:
+            result = strat(action, self)
+            if result is False:
+                break  # coverage: ignore
+            if result is True:
+                return True
+            assert result is NotImplemented, str(result)
+
+        return NotImplemented
+
+
+def _strat_act_on_stabilizer_ch_form_from_single_qubit_decompose(
+        val: Any, args: 'cirq.ActOnStabilizerCHFormArgs') -> bool:
+    if num_qubits(val) == 1:
+        if not has_unitary(val):
+            return NotImplemented
+        u = unitary(val)
+        clifford_gate = SingleQubitCliffordGate.from_unitary(u)
+        if clifford_gate is not None:
+            for axis, quarter_turns in clifford_gate.decompose_rotation():
+                if axis == pauli_gates.X:
+                    if common_gates.XPowGate(exponent=quarter_turns /
+                                             2)._act_on_(args) is not True:
+                        return False  # coverage: ignore
```

Comment:
I'm really worried about these coverage: ignore-s all around - this should be only the case if it is impossible to cover.

Reply:
Removed them all